runtime-sdk/src/modules/rofl/mod.rs

```diff
@@ -422,18 +446,75 @@ impl<Cfg: Config> Module<Cfg> {
         Ok(types::DeriveKeyResponse { key })
     }
 
+    fn derive_app_key_id(
+        app: &app_id::AppId,
+        kind: types::KeyKind,
+        scope: types::KeyScope,
+        key_id: &[u8],
+        reg: Option<types::Registration>,
+    ) -> Result<keymanager::KeyPairId, Error> {
+        // Build the base key identifier.
+        //
+        // We use the following tuple elements which are fed into TupleHash to derive the final key
+        // identifier, in order:
+        //
+        // - V1 context domain separator.
+        // - App ID.
+        // - Encoded kind.
+        // - Key ID.
+        // - Optional CBOR-serialized extra domain separation.
+        //
+        let kind_id = &[kind as u8];
+        let mut key_id = vec![ROFL_DERIVE_KEY_CONTEXT, app.as_ref(), kind_id, key_id];
+        let mut extra_dom: BTreeMap<&str, Vec<u8>> = BTreeMap::new();
+
+        match scope {
+            types::KeyScope::Global => {
+                // Nothing to do here, global keys don't include an explicit scope for backwards
+                // compatibility.
+            }
+            types::KeyScope::Node => {
+                // Fetch node identifier corresponding to the application instance.
+                let node_id = reg.ok_or(Error::InvalidArgument)?.node_id;
+
+                extra_dom.insert("scope", [scope as u8].to_vec());
+                extra_dom.insert("node_id", node_id.as_ref().to_vec());
+            }
+            types::KeyScope::Entity => {
+                // Fetch entity identifier corresponding to the application instance.
+                let entity_id = reg
+                    .ok_or(Error::InvalidArgument)?
+                    .entity_id
+                    .ok_or(Error::InvalidArgument)?;
+
+                extra_dom.insert("scope", [scope as u8].to_vec());
+                extra_dom.insert("entity_id", entity_id.as_ref().to_vec());
+            }
+        };
+
+        // Add optional extra domain separation.
+        let extra_dom = if !extra_dom.is_empty() {
+            cbor::to_vec(extra_dom)
+        } else {
+            vec![]
```

Comment:
Can this be removed, i.e. if `extra_dom` is not empty, we cbor the data and push it to `key_id`?